Core/include/Acts/Definitions/Direction.hpp

```diff
@@ -0,0 +1,174 @@
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/Algebra.hpp"
+
+#include <iosfwd>
+#include <string>
+
+namespace Acts {
+
+/// The direction is always with respect to a given momentum, surface normal or
+/// other general axes
+class Direction final {
+ private:
+  enum class Value : int {
+    Negative = -1,
+    Positive = 1,
+  };
+
+ public:
+  static constexpr auto Negative = Value::Negative;
+  static constexpr auto Positive = Value::Positive;
+
+  static constexpr auto Backward = Value::Negative;
+  static constexpr auto Forward = Value::Positive;
+
+  /// This turns a signed value into a direction
+  ///
+  /// @param scalar is the signed value
+  ///
+  /// @return a direciton enum
+  static inline constexpr Direction fromScalar(ActsScalar scalar) {
+    assert(scalar != 0);
+    return scalar > 0 ? Value::Positive : Value::Negative;
+  }
+
+  /// Convert and index [0,1] to a direction e.g. for sorting in
+  /// std::array<T, 2u>
+  ///
+  /// @param index is the direction at input
+  static inline constexpr Direction fromIndex(std::size_t index) {
+    if (index == 0u) {
+      return Value::Negative;
+    }
+    return Value::Positive;
+  }
+
+  /// Convert dir to index [0,1] which allows to store direction dependent
+  /// objects in std::array<T, 2u>
+  ///
+  /// @return either 0 or 1
+  static inline constexpr std::size_t index(Direction dir) {
+    return dir.index();
+  }
+
+  /// Invert a direction
+  ///
+  /// @param dir is the direction at input
+  ///
+  /// @return an opposite direction
+  static inline constexpr Direction invert(Direction dir) {
+    return dir.invert();
+  }
+
+  /// Convert dir to index [0,1] which allows to store direction dependent
+  /// objects in std::array<T, 2u>
+  ///
+  /// @return either 0 or 1
+  inline constexpr std::size_t index() const {
+    if (m_value == Value::Negative) {
+      return 0u;
+    }
+    return 1u;
+  }
+
+  /// Turns the direction into a signed value
+  ///
+  /// @return a signed value
+  inline constexpr int sign() const { return static_cast<int>(m_value); }
+
+  /// Reverse the direction
+  ///
+  /// @param dir is the direction at input
+  ///
+  /// @return an opposite direction
+  inline constexpr Direction invert() const {
+    return (m_value == Value::Positive) ? Value::Negative : Value::Positive;
+  }
+
+  std::string toString() const;
+
+  inline constexpr Direction() = default;
+  inline constexpr Direction(Value value) : m_value(value) {}
+
+  inline constexpr bool operator==(Direction other) const {
+    return m_value == other.m_value;
+  }
+
+  inline constexpr bool operator!=(Direction other) const {
+    return m_value != other.m_value;
+  }
+
+ private:
+  Value m_value = Value::Positive;
+};
+
+std::ostream& operator<<(std::ostream& os, Direction dir);
+
+// Direction * T
+
+inline constexpr auto operator*(Direction dir, int value) {
```

Comment:
I think in these cases I'd probably prefer explicit return types, rather than `auto`.